avoid_list now supports <s> in negative constraints

mjpost opened the pull request:
Simple fix that lets you add `<s>` as the beginning of a sequence to avoid.

#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [x] Were system tests modified? If so did you run these at least 5 times to account for the variation across runs?
- [x] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./style-check.sh`)
- [x] You have considered writing a test
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

fhieber commented on contrib/sacrebleu/sacrebleu.py:
```diff
@@ -37,7 +37,7 @@
 import unicodedata
 
-VERSION = '1.3.0'
+VERSION = '1.2.10'
```
not sure if sneaking this in is the right way. Why not make a small separate PR and explain the downgrade?

fhieber commented on sockeye/inference.py:
```diff
@@ -1461,6 +1461,7 @@ def _beam_search(self,
             avoid_states = constrained.AvoidBatch(self.batch_size, self.beam_size,
                                                   avoid_list=raw_avoid_list,
                                                   global_avoid_trie=self.global_avoid_trie)
+            avoid_states.consume(best_word_indices)
```
any chance for a test? :)

mjpost replied:
I can't sneak anything by you! Okay, I'll add one.

tdomhan commented:
what does it mean to allow to avoid `<s>`?

tdomhan commented:
is the idea to have an avoid phrase to only apply at the beginning of the sentence?

fhieber commented:
I would think that this enables users to say that the translation should not start with "I think", when specifying an avoid constraint as such: "<s> I think".

mjpost commented:
Yes, exactly. This allows you to avoid a prefix, e.g., "<s> I think". Trying to specify "<s>" has no effect.

tdomhan commented:
Thanks! Could you add that explanation to the CHANGELOG?

mjpost commented:
Okay, I added test cases and refactored them in the process. Adding the test cases uncovered a bug.

This had to be an integration test since it is testing the output of `_beam_search()`. But the sequence copy integration test was already getting unwieldy, and was never ideal for testing constraints. The refactored tests (a) remove testing constraints from `test_seq_copy_int.py` and move it into `test_constraints_int.py`. There, we first decode in an unconstrained setting, then add positive constraints, and negative constraints, and ensure that the proper thing happens in the output.

This uncovered a bug. Say you have a negative constraint of "6 5 5". This wouldn't prevent the output "6 6 5 5" from being generated, because the state wasn't getting properly reset when generating the second "6". The code was resetting the state to the beginning of the trie, but actually, it needed to be taking the first step from the root of the trie.

fhieber commented on CHANGELOG.md:
```diff
@@ -10,6 +10,12 @@ Note that Sockeye has checks in place to not translate with an old model that wa
 
+## [1.18.43]
+### Fixed
+- `<s>` now supported as the first token in a multi-word negative constraint
```
[nit] could make this a bullet point and the item in 1.18.42 as well?

mjpost replied:
Sorry, I'm not sure I understand this one. This improvement doesn't apply to 1.18.42, right?

fhieber replied:
I should be more clear: the description is missing a dash "-", so does the previous entry in the change log (length calculation).

mjpost replied:
Where is the dash missing? This is what I see right now:

```
## [1.18.43]
### Added
- `<s>` now supported as the first token in a multi-word negative constraint
  (e.g., '<s> I think' to prevent a sentence from starting with 'I think')
### Fixed
- Bugfix in resetting the state of a multiple-word negative constraint

## [1.18.42]
### Changed
- Simplified gluon blocks for length calculation
```

fhieber commented on sockeye/lexical_constraints.py:
```diff
@@ -116,10 +116,21 @@ def consume(self, word_id: int) -> 'AvoidState':
+        (1) If the word is found in our set of outgoing child arcs, we take that transition.
+        (2) If the word is not found, and we are not in the root state, we need to reset.
+            This means we pretend we were in the root state, and see if we can take a step
+        (3) Otherwise, if we are not already the root state (i.e., we were partially through
```
"if we are not already IN the root state"

fhieber commented on sockeye/lexical_constraints.py:
```diff
@@ -116,10 +116,21 @@ def consume(self, word_id: int) -> 'AvoidState':
+        (2) If the word is not found, and we are not in the root state, we need to reset.
+            This means we pretend we were in the root state, and see if we can take a step
+        (3) Otherwise, if we are not already the root state (i.e., we were partially through
+            the trie), we need to create a new object to track the state, in the root state
```
"create a new object in root state" maybe?

fhieber commented on test/common.py:
```diff
@@ -375,6 +347,61 @@ def run_train_translate(train_params: str,
+                with patch.object(sys, "argv", params.split()):
+                    sockeye.translate.main()
+
+                print('JSON_INPUT', open(new_test_source_path).readlines(),
```
could we change this to a logging statement? with pytest, logging statements are easier to track than stdout output.

mjpost replied:
Oh, this should be removed entirely.

fhieber commented on test/common.py:
```diff
@@ -375,6 +347,61 @@ def run_train_translate(train_params: str,
+                      'UNCONSTRAINED', open(out_path).readlines(), sep='\n')
+                for json_input, constrained_out, unconstrained_out in zip(open(new_test_source_path), open(out_path_constrained), open(out_path)):
+                    jobj = json.loads(json_input)
+                    print('JSON_INPUT', json_input, 'CONSTRAINED', constrained_out, 'UNCONSTRAINED', unconstrained_out, sep='\n')
```
same here

fhieber commented on test/common.py:
```diff
@@ -375,6 +347,61 @@ def run_train_translate(train_params: str,
+                    jobj = json.loads(json_input)
+                    print('JSON_INPUT', json_input, 'CONSTRAINED', constrained_out, 'UNCONSTRAINED', unconstrained_out, sep='\n')
+                    if jobj.get(constraint_type, None) == None:
+                        assert constrained_out == unconstrained_out
```
would it be helpful to add error messages to these `assert` statements?

mjpost replied:
I can add comments, but none of the other asserts have comments, and I'm not sure exactly what to put...

fhieber replied:
lets leave as is, was just wondering about whether the if cases are clear, and/or what we are asserting. but after a second look its clear to me.

mjpost commented:
Running system tests, shouldn't be affected, will wait to merge till they finish.

mjpost commented:
System tests pass.
